python/hail/linalg/blockmatrix.py

```diff
@@ -553,103 +604,207 @@ def __neg__(self):
         op = getattr(self._jbm, "unary_$minus")
         return BlockMatrix(op())
 
-    @typecheck_method(b=oneof(numeric, block_matrix_type))
+    def _check_common_dim(self, b, op):
+        if self.shape[0] != b.shape[0] and self.shape[1] != b.shape[1]:
+            raise ValueError(f'incompatible shapes for {op}: {self.shape} and {b.shape}')
+
+    def _promote(self, b, op):
+        a = self
```

Comment:
What's the invariants for the return type?  Looking at it:

1. if b is SCALAR, it is a float,
1.5. if b is an ndarray, it is 2d (and (n,) => (1, n))
2. a is a block matrix,
3. a_form >= b_form

Is that it?

I would organize this to:

 - enforce (1.5),
 - compute forms of each, and make Form.of take the object, and have it handle all three kinds of objects,
 - enforce (3), but convert b to a block matrix so (2) is maintained,
 - make scalar b a float,
 - check dimensions,
 - (2) is maintained by construction.

I think this will make the logic a bit cleaner.

Reply:
The conversion applied to each object depends on both forms.

Reply:
Latest commit tries to make this clearer